context/app/static/js/components/bulkDownload/hooks.ts

```diff
@@ -0,0 +1,196 @@
+    size: 1000,
+  };
+  const { searchHits, isLoading } = useSearchHits<BulkDownloadDataset>(datasetQuery);
+  const datasets = searchHits.map(({ _source }) => _source);
+
+  // Which options to show in the dialog
+  const downloadOptions = ALL_BULK_DOWNLOAD_OPTIONS.map((option) => ({
+    ...option,
+    count: datasets.filter((dataset) => option.isIncluded(dataset)).length,
+  })).filter((option) => option.count > 0);
+
+  // Remove selected uuids from the list and deselect them in the table if needed
+  const removeUuidsOrRows = useCallback(
+    (uuidsToRemove: string[]) => {
+      if (deselectRows) {
+        deselectRows(uuidsToRemove);
+      }
+      setUuids(new Set([...uuids].filter((uuid) => !uuidsToRemove.includes(uuid))));
+    },
+    [deselectRows, setUuids, uuids],
+  );
+
+  const protectedDatasetsFields = useProtectedDatasetsForm({
+    selectedRows: new Set(uuids),
+    deselectRows: removeUuidsOrRows,
+    protectedDatasetsErrorMessage: (protectedDatasets) =>
+      `You have selected ${protectedDatasets.length} protected datasets.`,
+    trackEventHelper: (numProtectedDatasets) => {
+      trackEvent({
+        category: 'Bulk Download',
+        action: 'Protected datasets selected',
+        value: numProtectedDatasets,
+      });
+    },
+  });
+
+  const downloadMetadata = useCallback(
+    (datasetsToDownload: BulkDownloadDataset[]) => {
+      postAndDownloadFile({
+        url: '/metadata/v0/datasets.tsv',
+        body: { uuids: datasetsToDownload.map((dataset) => dataset.uuid) },
+        fileName: 'metadata.tsv',
+      })
+        .then(() => {
+          toastSuccessDownloadFile('Metadata');
+        })
+        .catch((e) => {
+          toastErrorDownloadFile('Metadata', () => downloadMetadata(datasetsToDownload));
+          console.error(e);
+        });
+    },
+    [toastSuccessDownloadFile, toastErrorDownloadFile],
+  );
+
+  const downloadManifest = useCallback(
+    (datasetsToDownload: BulkDownloadDataset[]) => {
+      const url = createDownloadUrl(
+        `${datasetsToDownload.map((dataset) => dataset.hubmap_id).join(' /\n')} /`,
+        'text/plain',
+      );
+
+      checkAndDownloadFile({ url, fileName: 'manifest.txt' })
+        .then(() => {
+          toastSuccessDownloadFile('Manifest');
+        })
+        .catch((e) => {
+          toastErrorDownloadFile('Manifest', () => downloadManifest(datasetsToDownload));
+          console.error(e);
+        });
+    },
+    [toastSuccessDownloadFile, toastErrorDownloadFile],
+  );
+
+  const handleClose = useCallback(() => {
+    reset();
+    close();
+  }, [close, reset]);
+
+  const onSubmit = useCallback(
+    ({ bulkDownloadOptions, bulkDownloadMetadata }: BulkDownloadFormTypes) => {
+      const datasetsToDownload = datasets.filter((dataset) =>
+        bulkDownloadOptions.some((option) =>
+          ALL_BULK_DOWNLOAD_OPTIONS.find(({ key }) => key === option)?.isIncluded(dataset),
+        ),
+      );
```

Comment:
This could also be avoided by separately requesting each option? Or maybe at the least we can iterate over the datasets once to separate them to avoid doing it here and counting them above?

Reply:
Good idea!